File: src/main/java/ladanse/website/domain/SignUp.java
package ladanse.website.domain;

import ladanse.website.domain.master.RoleType;
import ladanse.website.domain.master.SignUpType;

import javax.persistence.*;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "SignUp")
public class SignUp {
    private Long id;
    private Date since;
    private Date end;
    private Event event;
    private Account account;
    private SignUpType signUpType;
    private Set<ForRole> forRoles;

    protected SignUp() {
    }

    public SignUp(Account account, Event event, SignUpType signUpType, Set<RoleType> roleTypes) {
        setSignUpType(signUpType);
        setEvent(event);
        setAccount(account);
        setSince(new Date());

        Set<ForRole> roles = new HashSet<>();

        for (RoleType roleType : roleTypes) {
            ForRole forRole = new ForRole(this, roleType);

            roles.add(forRole);
        }

        setForRoles(roles);
    }

    public void cancel() {
        setEnd(new Date());
    }

    @Id
    @Column(name = "id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long getId() {
        return id;
    }

    protected void setId(Long id) {
        this.id = id;
    }

    @Basic
    @Column(name = "since")
    @Temporal(TemporalType.TIMESTAMP)
    public Date getSince() {
        return since;
    }

    protected void setSince(Date since) {
        this.since = since;
    }

    @Basic
    @Column(name = "end")
    @Temporal(TemporalType.TIMESTAMP)
    public Date getEnd() {
        return end;
    }

    protected void setEnd(Date end) {
        this.end = end;
    }

    @ManyToOne
    @JoinColumn(name = "eventId", nullable = false)
    public Event getEvent() {
        return event;
    }

    protected void setEvent(Event event) {
        this.event = event;
    }

    @ManyToOne
    @JoinColumn(name = "accountId", nullable = false)
    public Account getAccount() {
        return account;
    }

    protected void setAccount(Account account) {
        this.account = account;
    }

    @Basic
    @Column(name = "signUpType")
    @Enumerated(EnumType.STRING)
    public SignUpType getSignUpType() {
        return signUpType;
    }

    protected void setSignUpType(SignUpType signUpType) {
        this.signUpType = signUpType;
    }

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "signUp")
    public Set<ForRole> getForRoles() {
        return forRoles;
    }

    protected void setForRoles(Set<ForRole> forRoles) {
        this.forRoles = forRoles;
    }
}

File: src/LaDanse/DomainBundle/Entity/SignUp.php
<?php
/**
 * @license  http://opensource.org/licenses/gpl-license.php GNU Public License
 * @link     https://github.com/bderidder/ldm-guild-website
 */

namespace LaDanse\DomainBundle\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity
 * @ORM\Table(name="SignUp")
 */
class SignUp
{
    const REPOSITORY = 'LaDanseDomainBundle:SignUp';

    /**
     * @ORM\Column(type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    protected $id;

    /**
     * @ORM\ManyToOne(targetEntity="Event", inversedBy="signUps")
     * @ORM\JoinColumn(name="eventId", referencedColumnName="id", nullable=false)
     */
    protected $event;

    /**
     * @ORM\ManyToOne(targetEntity="Account")
     * @ORM\JoinColumn(name="accountId", referencedColumnName="id", nullable=false)
     */
    protected $account;

    /**
     * @ORM\Column(type="string", length=15, nullable=false)
     */
    protected $type;

    /**
     * @ORM\OneToMany(targetEntity="ForRole", mappedBy="signUp", cascade={"persist", "remove"})
     * @var $roles ArrayCollection
     */
    protected $roles;

    /**
     * Constructor
     */
    public function __construct()
    {
        $this->roles = new ArrayCollection();
    }

    /**
     * Get id
     *
     * @return integer 
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * Set event
     *
     * @param Event $event
     * @return SignUp
     */
    public function setEvent(Event $event = null)
    {
        $this->event = $event;

        return $this;
    }

    /**
     * Get event
     *
     * @return Event
     */
    public function getEvent()
    {
        return $this->event;
    }

    /**
     * Set account
     *
     * @param Account $account
     * @return SignUp
     */
    public function setAccount(Account $account = null)
    {
        $this->account = $account;

        return $this;
    }

    /**
     * Get account
     *
     * @return Account
     */
    public function getAccount()
    {
        return $this->account;
    }

    /**
     * Add roles
     *
     * @param ForRole $roles
     * @return SignUp
     */
    public function addRole(ForRole $roles)
    {
        $this->roles[] = $roles;

        return $this;
    }

    /**
     * Remove roles
     *
     * @param ForRole $roles
     */
    public function removeRole(ForRole $roles)
    {
        $this->roles->removeElement($roles);
    }

    /**
     * Get roles
     *
     * @return \Doctrine\Common\Collections\Collection 
     */
    public function getRoles()
    {
        return $this->roles;
    }

    /**
     * Set type
     *
     * @param string $type
     * @return SignUp
     */
    public function setType($type)
    {
        $this->type = $type;

        return $this;
    }

    /**
     * Get type
     *
     * @return string 
     */
    public function getType()
    {
        return $this->type;
    }

    public function toJson()
    {
        $simpleRoles = array();

        for($i = 0; $i < $this->roles->count(); $i++)
        {
            $simpleRoles[] = $this->roles->get($i)->getRole();
        }

        return (object) array(
            'signUpId' => $this->id,
            'type'     => $this->type,
            'roles'    => $simpleRoles
        );
    }
}
